File: packages/guards/src/pii-ru.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiRuEntity = 'inn' | 'snils' | 'passport-ru' | 'phone-ru';
interface PiiRuOptions { entities: PiiRuEntity[]; action: 'block' | 'warn' | 'mask'; }
interface PiiRuMatch { type: PiiRuEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiRuEntity, RegExp> = {
  inn: /\b\d{10,12}\b/g,
  snils: /\b\d{3}-\d{3}-\d{3}\s?\d{2}\b/g,
  'passport-ru': /\b\d{2}\s?\d{2}\s?\d{6}\b/g,
  'phone-ru': /\b(?:\+?7|8)[-.\s]?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{2}[-.\s]?\d{2}\b/g,
};
const MASK: Record<PiiRuEntity, string> = { inn: '[ИНН]', snils: '[СНИЛС]', 'passport-ru': '[Паспорт]', 'phone-ru': '[Телефон]' };

function detect(text: string, entities: PiiRuEntity[]): PiiRuMatch[] {
  const m: PiiRuMatch[] = [];
  for (const e of entities) { const re = new RegExp(PATTERNS[e].source, 'g'); let x; while ((x = re.exec(text))) m.push({ type: e, value: x[0], start: x.index, end: x.index + x[0].length }); }
  return m.sort((a, b) => b.start - a.start);
}

export function piiRu(options: PiiRuOptions): Guard {
  return { name: 'pii-ru', version: '0.1.0', description: 'Russian PII (INN, SNILS, passport, phone)', category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matches = detect(text, options.entities); const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-ru', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') { let r = text; for (const m of matches) r = r.slice(0, m.start) + MASK[m.type] + r.slice(m.end); return { guardName: 'pii-ru', passed: true, action: 'override', overrideText: r, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type }) => ({ type })) } }; }
      return { guardName: 'pii-ru', passed: false, action: options.action, message: `Russian PII: ${[...new Set(matches.map(m => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_ru.py
"""Russian PII detection and masking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "passport-ru": re.compile(r"\b\d{2}\s?\d{2}\s?\d{6}\b"),
    "inn": re.compile(r"\b\d{10}(?:\d{2})?\b"),
    "snils": re.compile(r"\b\d{3}-\d{3}-\d{3}\s?\d{2}\b"),
    "phone-ru": re.compile(r"\b(?:\+7|8)\s?\(?\d{3}\)?\s?\d{3}[\s\-]?\d{2}[\s\-]?\d{2}\b"),
}

_MASK_LABELS = {"passport-ru": "[ПАСПОРТ]", "inn": "[ИНН]", "snils": "[СНИЛС]", "phone-ru": "[ТЕЛЕФОН]"}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches = []
    for e in entities:
        pat = _PATTERNS.get(e)
        if pat:
            for m in pat.finditer(text):
                matches.append({"type": e, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    r = text
    for m in matches:
        r = r[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + r[m["end"] :]
    return r


class _PiiRu:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-ru"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-ru", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-ru", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-ru", passed=False, action=self.action, message="Russian PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_ru(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiRu:
    return _PiiRu(action=action, entities=entities)
